### ISSUE TITLE
Transferring Pokemon even with keepMinDuplicate = 1

### ISSUE THREAD

**Issue by CreativiTimothy** [NoxxDev/NecroBot#702]
### Expected Behavior
Keeps all Pokemon with 1 duplicate max except 95+ IV Pokemon or Pokemon with high enough CP specified in PokemonTransferConfig.

### Actual Behavior
Transfers every single Pokemon except favorited ones (which are impossible to transfer anyway)
Transferred my 2364CP Lapras (it was my only Lapras, so after it transferred, I had 0 Laprases left). Screenshot: http://prntscr.com/cbx54l

### Your config.json
    "TransferWeakPokemon": true,
    "TransferDuplicatePokemon": true,
    "TransferDuplicatePokemonOnCapture": true,
    "RenamePokemon": true,
    "RenameOnlyAboveIv": true,
    "RenameTemplate": "{1}{0}",
    "FavoriteMinIvPercentage": 95.0,
    "AutoFavoritePokemon": true,
    "EvolveAllPokemonWithEnoughCandy": true,
    "EvolveKeptPokemonsAtStorageUsagePercentage": 99.0,
    "KeepPokemonsThatCanEvolve": true,
    "KeepMinCp": 5000,
    "KeepMinIvPercentage": 95.0,
    "KeepMinLvl": 6,
    "KeepMinOperator": "or",
    "UseKeepMinLvl": false,
    "PrioritizeIvOverCp": false,
    "KeepMinDuplicatePokemon": 1,

  "PokemonsTransferFilter": {
    "Lapras": {
      "KeepMinCp": 2235,
      "KeepMinLvl": 25,
      "UseKeepMinLvl": false,
      "KeepMinIvPercentage": 95.0,
      "KeepMinDuplicatePokemon": 1,
      "Moves": [
        []
      ],
      "DeprecatedMoves": [],
      "KeepMinOperator": "or",
      "MovesOperator": "or"
    },
}

### Other Information
Version: 0.9.6
Windows Version:
What is affected: transfer

**Comment by CreativiTimothy** [comment]
Is TransferWeakPokemon = true causing the problem? For some reason, it also transfers my 100IV Pokemon even though I set it to auto-favorite 95+ IV Pokemon

**Comment by kuntau** [comment]
I had an anecdotal experience TransferWeakPokemon messed with my high CP pokemon but no highest IV, never set it true ever again

**Comment by Skolzack** [comment]
Necro transfer favorited pokemons? LUL

**Comment by Skolzack** [comment]
set "PrioritizeIvOverCp" to true and see if it happens again

**Comment by CreativiTimothy** [comment]
@Skolzack 

> Transfers every single Pokemon except favorited ones (which are impossible to transfer anyway)

**Comment by iwean** [comment]
Your set >>> "KeepMinCp": 5000,
so the bot think CP 2364 is Weak :smile:

**Comment by CreativiTimothy** [comment]
@iwean PokemonTransferFilter should override global KeepMinCP: 5000. otherwise, PokemonTransferFilter would be useless

**Comment by iwean** [comment]
Yup... but TransferWeakPokemon override everyting... base on global config (KeepMinCP: 5000)

**Comment by CreativiTimothy** [comment]
@iwean Yeah, apparently. TransferWeakPokemon /should/ follow the PokemonTransferFilter first before global settings. All the hard work into configuring the PokemonTransferFilter becomes useless when TransferWeakPokemon is on, so TransferWeakPokemon must be fixed

**Comment by iwean** [comment]
Nope... it's a simple mode you know? If you set it to true.... you must not set KeepMin or TransferFilter.
Everytime bot catch the strongger.... the weak transfered.
That's how simple mode do. :smile:

**Comment by CreativiTimothy** [comment]
@iwean ...
you don't understand.
expected behavior may be different among other people but generally most people will agree on a consensus on what it is expected. kuntau agrees with me. Let's see what NoxxDev thinks

**Comment by iwean** [comment]
I do understand what you excpected.... please set TransferWeak to false... because you want everyting as you expect by setting on KeepMin & TransferFilter....

**Comment by kuntau** [comment]
TransferWeakPokemon even ignore at least 1 of the same pokemon, and
transfer the only putyourrarepokemonhere that you have

**Comment by iwean** [comment]
@kuntau Yeah.... putyourrarepokemonhere that you have... the stronger one :smile:

**Comment by CreativiTimothy** [comment]
@kuntau Yeah. I think that part is intentional though. It will ignore the duplicate option because the purpose of that transferweakpokemon is to make up space. however, i dont think the ignoring of pokemontransferfilter is on purpose

**Comment by iwean** [comment]
As I said... it's a simple mode (set true to TransferWeak) if you lazy to set everyting else. I mean by set KeepMin and TransferFilter.

**Comment by CreativiTimothy** [comment]
TransferWeakPokemon: true
AutoFavoriteIV: 95
KeepMinIV: 95

*Transfers 100IV 2700CP Dragonite*

AGHHHHH!! It ain't weak

**Comment by CreativiTimothy** [comment]
@iwean The wiki says this:

TransferWeakPokemon (Value)

Transfer Pokemon that do not meet your minimum CP, IV, or Level values, depending on your configured priorities.




My dragonite DOES meet the requirement

**Comment by iwean** [comment]
2700 is weak than KeepMin 5000 :smile:

**Comment by CreativiTimothy** [comment]
@iwean keepmincp: 5000

OR <------

keepminiv: 95

keep pokemon if EITHER 5000CP+ or 95IV+

**Comment by iwean** [comment]
TransferWeak prioritize on CP.... only if youset true to PrioritizeIvOverCp... it's prioritiza IV.

**Comment by iwean** [comment]
Just set it to FALSE.... the TransferWeak :smile:

**Comment by CreativiTimothy** [comment]
@iwean I don't think you're a native English speaker

**Comment by iwean** [comment]
Yup... sorry for my bad English :dancer:

**Comment by kalamera** [comment]
`"KeepMinCp": 5000`

Just LoL :)

imho: 

> "TransferWeakPokemon" is much better than the other two "TransferDuplicate..." options
> You should never set all the transfer to "True". You need to pick which "Transfer..." you will use.
> The "TransferWeakPokemon" just works better if you set the values right.

**Comment by iwean** [comment]
@kalamera -  agree :+1: 
The simple and the better ways :smile:

**Comment by NeiPCs** [comment]
Well,

After read all drama and check at source code of bot on TransferWeakPokemonTask I have something to say:
@CreativiTimothy You are right about you expect from this option.
@iwean Sorry, but "Just set it to FALSE" is just the lazy way to avoid feature, by ignoring the robust mechanism of bot and getting satisfected with the "simple basic things done". Keep going, but this section is not for you, just ignore this issue and SET IT TO FALSE in your head.

The source code compare pokemon stats with config file before do things:

- Filter pokemons NOT IN PokemonsNotToTransfer and order by CP (from weak to high)
- If KeepPokemonsThatCanEvolve = true, then Filter again all NOT IN PokemonsToEvolve
- Order again by CP (from weak to high)
- Check every pokemon in the filtered list by follow conditions:
  - Pokemon CP >= KeepMinCp 
  - OR
  - (If PrioritizeIvOverCp = true, Pokemon IV >= KeepMinIvPercentage)
  - OR
  - (If UseKeepMinLvl = true, Pokemon Level >= KeepMinLvl)
  - OR
  - Pokemon is Favorite

If any of above conditions is TRUE, then bot will KEEP and NOT transfer the pokemon.

But something is missing in there:

 - The bot don't check **KeepMinOperator**, and use OR as default
 - The bot don't check **PrioritizeIvOverCp**, and use CP as default parameter to check if pokemon is weak

**Comment by Skolzack** [comment]
@NeiPCs > The bot don't check PrioritizeIvOverCp, and use CP as default parameter to check if pokemon is weak

That's not true, when you set "PrioritizeIvOverCP" to true he will always check the IV first.. The real problem is that "KeepMinCP": 5000 of the guy above that are messing everything around.

**Comment by NeiPCs** [comment]
Please don't generalize, I'm talking about ONE Function of bot: **TransferWeakPokemonTask**

**Comment by iwean** [comment]
And I'm talking about lazy thing function.... :smile: 
For me <<< I'm set it to true of course :ROTFL:

**Comment by iwean** [comment]
And keep on your mind, my dear @NeiPCs .... if you set KeepMinCP, that's the base for TransferWeakPokemonTask to compare weak point.

But if you set PrioritizeIvOverCp to TRUE, the weak point to compare change from CP to IV.

**Comment by NeiPCs** [comment]
@iwean In the pratice it's not implemented at all (and yeat) 
Maybe we need someone to make some TODO in source code.

**Comment by kuntau** [comment]
@iwean @Skolzack that is not the problem, if the function functioning as expected with `KeepMinOperator = or` operation. I can have 10CP 100IV Weedle and it shouldn't be transfered

**Comment by Skolzack** [comment]
@kuntau Dude, of course it's the problem... cmon lol

**Comment by Yinyinka** [comment]
Dont know where you messed with your config but this is mine and it is working fine. Catch all, keep PokemonWhichCanEvolve, keep IV 95+ and CP 1999+ and transfer the rest. I have many IV95+ pokemon with CP under 1999 and they are still in my pocket. :P
    "KeepMinCp": 1999,
    "KeepMinIvPercentage": 95.0,
    "KeepMinLvl": 1,
    "KeepMinOperator": "or",
    "UseKeepMinLvl": false,
    "PrioritizeIvOverCp": true,
    "KeepMinDuplicatePokemon": 0,

**Comment by wilsondgonzales** [comment]
90% of the people here are misunderstanding @CreativiTimothy . Regardles of how he sets his KeepMinCP, the function TransferWeakPokemon should not execute the way it's executing. THAT is what the issue he's trying to point out is. People should stop pointing at KeepMinCP variable because it is not the problem he is having. Looking at it throughly, his IV parameter should pass his dragonite as "Not Weak" but it doesn't get excluded by the function of TransferWeakPokemon. That is the issue he's bringing to attention

**Comment by CreativiTimothy** [comment]
@wilsondgonzales THANK YOU!

**Comment by langit7** [comment]
Low CP mean weak,  for me that use Google translate it's not hard to understand,  10 ratatta with 100iv and 10cp will not win  again  dragonite  with, 10iv and 2k cp

**Comment by Brazuca87** [comment]
lmao... this is almost the exact discussion I've had on #737... I believe this iwean guy does not understand English, or doesn't know what he is saying... On that topic he claims I should turn this TransferWeakPokemon to true, so that issue won't happen, and then here he says, basically, if it was turned on, the other pokemon would be transferred instead (or actually, both would!)

Non-sense... I'll see if I fix the code for my thread tonight... If possible, I'll take a peak at this one as well... My coding skills might be a bit rusty, so might take a while :)

**Comment by iwean** [comment]
:ROTFL: @Brazuca87 
Even I'm not understand English... I still keep my 100% IV on my inventory
....
![screenshot_160](https://cloud.githubusercontent.com/assets/4365551/18156493/ce0492a2-7040-11e6-8f9f-d23f4dac5d7d.jpg)

**Comment by iwean** [comment]
And this is what will transfered when my inventory FULL.... You can see there's nothing special with my understanding English setting of config.json. :LOL:
...
![screenshot_161](https://cloud.githubusercontent.com/assets/4365551/18156745/b94cc1de-7042-11e6-9860-31b5c20e8bf6.jpg)

**Comment by CreativiTimothy** [comment]
@iwean inb4 TransferWeakPokemon = true
